p4-16/spec/P4-16-spec.mdk

```diff
@@ -6769,55 +6782,81 @@ and illegal specifications of default actions for the `table t`:
 
 While table entries are typically installed by the control plane,
 tables may also be initialized at compile time with a set of entries.
-This is useful in situations where tables are
+
+Declaring these entries with `const entries`
+is useful in situations where tables are
 used to implement fixed algorithms---defining table entries statically
 enables expressing these algorithms directly in P4,
 which allows the compiler to infer how the table is actually used and
 potentially make better allocation decisions for targets with limited
-resources. Entries declared in the P4 source are installed in the
+resources.
+
+Declaring entries with `entries` (without the `const` qualifier)
+enables one to specify a mix of some immutable entries that are always
+in the table, and some mutable entries that the control plane is
+allowed to later change or remove.
+
+Entries declared in the P4 source are installed in the
 table when the program is loaded onto the target. Entries cannot be
 specified for a table with no key (see Sec. [#sec-table-keys]).
 
 Table entries are defined using the following syntax:
 
 ~ Begin P4Grammar
 tableProperty
- : const ENTRIES '=' '{' entriesLlist '}' /* immutable entries */
+ : optAnnotations const ENTRIES '=' '{' entriesLlist '}' /* immutable entries */
+ | optAnnotations ENTRIES '=' '{' entriesList '}' /* mix of mutable and/or immutable entries */
 
 entriesList
  : /* empty */
  | entriesList entry
  ;
 
+optCONST
+ : /* empty */
+ | const
+ ;
+
+optEntryPriority
+ : /* empty */
+ | priority '=' INTEGER
+ ;
+
 entry
- : keysetExpression ':' actionRef optAnnotations ';'
+ : optCONST optEntryPriority keysetExpression ':' actionRef optAnnotations ';'
```

Comment:
Hmm, an entry would look like
`const priority = 1     2 ... 3   :   a(t)    @opn;`
I don't know if like 1 and 2 being next to each other there without some kind of separator. 
The P4 language has separators for most things, having a separator removes the ambiguous issue too and why the need () for the expression case too.
I am ok with not having a separator but many folks might have to do a double take of looking at the code to understand it.

Reply:
I agree that does not look great.

Note that the syntax also allows one to write it more clearly as:
```
const priority = 1 (2 .. 3) : a(t) @opn;
```
We could recommend that people use the parentheses there for clarity, perhaps?